### PR TITLE
For #377: Added InterceptAttack effect and tests

### LINKED ISSUE #377: Implemented Paladin Riot Shield card
Size: 2 Units

### PR DESCRIPTION
**Pull request by paulodamaso**
Closes #377: Added InterceptAttack effect and tests

### CODE REVIEW

File: src/ProjectNeon/Assets/Features/Effects/Attack.cs [hunk 1]
```diff
@@ -6,7 +6,7 @@
 public sealed class Attack 
 {
     public Member Attacker { get; }
-    public Member Target { get; }
+    public Member Target { get; set; }
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/Attack.cs]: Attacks must be `immutable`

File: src/ProjectNeon/Assets/Z_Tests/Features/Effects/InterceptAttackTest.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    [Test]
+    public void InterceptAttack_ApplyEffect_AttackIsIntercepted()
+    {
+        Member paladin = TestMembers.Create(s => s.With(StatType.MaxHP, 10).With(StatType.Damagability, 1f));
+        Member ally = TestMembers.With(StatType.MaxHP, 10);
+        Member attacker = TestMembers.With(StatType.Attack, 5);
+        Attack attack = new Attack(attacker, ally, 5);
+
+        AllEffects.Apply(new EffectData { EffectType = EffectType.InterceptAttack }, paladin, new MemberAsTarget(ally));
+        BattleEvent.Publish(attack);
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Z_Tests/Features/Effects/InterceptAttackTest.cs]: This needs to work correctly, even if there are multiple active subscribers to the Attack event. 

The current design doesn't account for this.

File: src/ProjectNeon/Assets/Features/Effects/InterceptAttack.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+class InterceptAttack : Effect
+{
+    private Member _performer;
+    private Target _effectTarget;
+
+    void Effect.Apply(Member source, Target target)
+    {
+        _performer = source;
+        _effectTarget = target;
+        BattleEvent.Subscribe<Attack>(attack => Execute(attack), this);
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/InterceptAttack.cs]: This must be cleaned up.

This effect is supposed to expire at the end of the turn.

File: src/ProjectNeon/Assets/Data/Cards/Paladin/RiotShield.asset [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  m_PrefabInstance: {fileID: 0}
+  m_PrefabAsset: {fileID: 0}
+  m_GameObject: {fileID: 0}
+  m_Enabled: 1
+  m_EditorHideFlags: 0
+  m_Script: {fileID: 11500000, guid: 6b41b1118d12e5c48a6e7ed020b96a84, type: 3}
+  m_Name: RiotShield
+  m_EditorClassIdentifier: 
+  art: {fileID: 21300000, guid: 29681aeecbd74e4439e488ec79235ed9, type: 3}
+  description: Intercept all attacks this round. +(Lvl) Armor this round.
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Data/Cards/Paladin/RiotShield.asset]: Text on card should match the `+2 Armor` of the design.

**Comment by paulodamaso** [comment]
@SilasReinagel I've updated the code, please take a look

File: src/ProjectNeon/Assets/Features/Effects/InterceptAttack.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+class InterceptAttack : Effect
+{
+    private Member _performer;
+    private Target _effectTarget;
+
+    void Effect.Apply(Member source, Target target)
+    {
+        _performer = source;
+        _effectTarget = target;
+        BattleEvent.Subscribe<AttackPerformed>(attackPerformed => Execute(attackPerformed.Attack), this);
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/InterceptAttack.cs]: This is too late. The damage has already been dealt to the original target.

File: src/ProjectNeon/Assets/Z_Tests/Features/Effects/InterceptAttackTests.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
+﻿
+using NUnit.Framework;
+using UnityEngine;
+
+public sealed class InterceptAttackTests
+{
+    [Test]
+    public void InterceptAttack_ApplyEffect_AttackIsIntercepted()
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Z_Tests/Features/Effects/InterceptAttackTests.cs]: This test doesn't check the health of the original target.

Comment (paulodamaso) [reply src/ProjectNeon/Assets/Z_Tests/Features/Effects/InterceptAttackTests.cs]: @SilasReinagel Fixed

**Comment by paulodamaso** [comment]
@SilasReinagel PLease take a look

File: src/ProjectNeon/Assets/Features/Effects/AllEffects.cs [hunk 7]
```diff
@@ -27,6 +27,7 @@ public static class AllEffects
         { EffectType.ShieldAttackedOnAttack, e => new EffectOnAttacked(new SimpleEffect((src, m) => m.GainShield(e.IntAmount * src.State.Toughness()))) },
         { EffectType.DamageAttackerOnAttack, e => new EffectOnAttacker(new SimpleEffect((src, m) => m.TakePhysicalDamage(e.IntAmount * src.State.Attack()))) },
         { EffectType.StealLifeNextAttack, e => new Recurrent(new StealLife(e.IntAmount), 1)},
+        { EffectType.InterceptAttackForTurns, e => new Recurrent(new InterceptAttack(), e.NumberOfTurns)},
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/AllEffects.cs]: Supposed to be for Number of Turns, not number of instances.

File: src/ProjectNeon/Assets/Features/Effects/AllEffects.cs [hunk 8]
```diff
@@ -27,6 +27,7 @@ public static class AllEffects
...
         { EffectType.Attack, e => new Attack(e.IntAmount)}
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/AllEffects.cs]: The damage amount can't be known yet. This should take the modifier.

File: src/ProjectNeon/Assets/Features/Effects/Attack.cs [hunk 9]
```diff
@@ -1,18 +1,18 @@
...
 
 public sealed class Attack  : Effect
 {
 
-    public Member Attacker { get; private set; }
-    public Target Target { get; private set; }
-    public int Damage { get; }
+    public Member Attacker { get; set; }
+    public Target Target { get; set; }
+    public int Damage { get; set; }
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/Attack.cs]: Damage amount can't be known until after the attack has resolved, or at least after Targets have been finalized.

File: src/ProjectNeon/Assets/Features/Effects/InterceptAttack.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    void Execute(Attack attack)
+    {
+        _effectTarget.Members.ForEach(
+            target => {
+                if (target.Equals(attack.Target.Members[0]))
+                {
+                    attack.Effect = new NoEffect();
+                    new Attack(attack.Damage).Apply(attack.Attacker, _performer);
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/InterceptAttack.cs]: Same problem with damage being passed to the `Attack`.

The amount will be calculated based on the Attacked and the Target stats.

File: src/ProjectNeon/Assets/Features/Effects/AllEffects.cs (hunk 7 above)
Comment (paulodamaso) [reply src/ProjectNeon/Assets/Features/Effects/AllEffects.cs]: @SilasReinagel Sorry, my bad. Fixed

**Pull request merged**